README.md
# Internet check

A small component to check if the internet is accessible. Original idea by [Stackoverflow #4860365](http://stackoverflow.com/questions/4860365/determine-in-php-script-if-connected-to-internet).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
# Usage

The very basic usage (with default parameters) looks like this:

```php
$available = \lfischer\internet\Internet::available();
```

But of course there are some options you can provide to handle the responses as you need them:

```php
use \lfischer\internet\Internet;
use \lfischer\internet\InternetException;
use \lfischer\internet\InternetProblemException;

// Check the availability by connecting to Google on port 80.
$available = (new Internet('www.google.com', 80))->check();

//
try {
    $internet = new Internet(
        'www.google.com', 
        80, 
        Internet::EXCEPTION_ON_UNAVAILABILITY + Internet::PROBLEM_AS_EXCEPTION
    );

    $available = $internet->check();
} catch (InternetException $e) {
    // The internet is not available.
    $internet->getErrorString();
    $internet->getErrorNumber();
    $e->getMessage();
} catch (InternetProblemException $e) {
    // There was a problem while checking the availability.
     $internet->getErrorString();
     $internet->getErrorNumber();
     $e->getMessage();
 }
```

## Options

You can pass some options to change the behaviour in case of problems.

- `Internet::EXCEPTION_ON_UNAVAILABILITY`\
Will throw an `InternetException` exception if the internet is unavailable (instead of returning `false`).
- `Internet::PROBLEM_AS_EXCEPTION`\
Will throw an `InternetProblemException` exception if the internet availability could not be checked due to a problem (instead of returning `false`).
- `Internet::PROBLEM_AS_TRUE`\
Will return `true` in case of a problem during the availability check since you'd like to assume something else went wrong.

# Static code analysis and code style

The code is being statically analyzed with the help of [vimeo/psalm](https://packagist.org/packages/vimeo/psalm). The PSR2 code style will be checked/applied with the help of [friendsofphp/php-cs-fixer](https://packagist.org/packages/friendsofphp/php-cs-fixer).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2020-01-23 
### Added
- Changelog

### Changed
- Changed PHP requirement to `>=7.1`

## [1.0.0] - 2020-01-21
### Added
- New class to check if the internet is currently available `\lfischer\internet\Internet`


src/Internet.php
<?php

namespace lfischer\internet;

use Exception;
use Psr\Log\LoggerInterface;

/**
 * Class Checker
 * @package lfischer\internet
 * @see http://stackoverflow.com/questions/4860365/determine-in-php-script-if-connected-to-internet
 */
class Internet
{
    /**
     * @var string
     */
    private $hostname = 'www.google.com';

    /**
     * @var int
     */
    private $port = 80;

    /**
     * @var float
     */
    private $timeout = 10.0;

    /**
     * @var int
     */
    private $options = self::PROBLEM_AS_TRUE;

    /**
     * @var LoggerInterface|null
     */
    private $logger = null;

    /**
     * @var int
     */
    private $errorNumber = 0;

    /**
     * @var string
     */
    private $errorString = '';

    /**
     * Options for error handling.
     */
    public const EXCEPTION_ON_UNAVAILABILITY = 1;
    public const PROBLEM_AS_EXCEPTION = 2;
    public const PROBLEM_AS_TRUE = 4;

    /**
     * Checker constructor.
     *
     * @param string|null $hostname
     * @param int|null $port
     * @param int|null $options
     * @param float|null $timeout
     * @param LoggerInterface|null $logger
     */
    public function __construct(string $hostname = null, int $port = null, int $options = null, float $timeout = null, LoggerInterface $logger = null)
    {
        if ($hostname !== null) {
            $this->hostname = $hostname;
        }

        if ($port !== null) {
            $this->port = $port;
        }

        if ($timeout !== null) {
            $this->timeout = $timeout;
        }

        if ($options !== null) {
            $this->options = $options;
        }

        if ($logger !== null) {
            $this->logger = $logger;
        }
    }

    /**
     * Get the error number of the socket connection.
     *
     * @return int
     */
    public function getErrorNumber(): int
    {
        return (int)$this->errorNumber;
    }

    /**
     * Get the error string of the socket connection.
     *
     * @return string
     */
    public function getErrorString(): string
    {
        return (string)$this->errorString;
    }

    /**
     * Check the connection with the provided options.
     *
     * @return bool
     * @throws InternetException
     * @throws InternetProblemException
     */
    public function check(): bool
    {
        try {
            $connected = @fsockopen($this->hostname, $this->port, $this->errorNumber, $this->errorString, $this->timeout);

            if ($connected) {
                fclose($connected);

                return true;
            }

            if ($this->logger !== null) {
                $this->logger->alert('Internet not available: ' . $this->errorNumber . ' - ' . $this->errorString);
            }

            if ($this->options & self::EXCEPTION_ON_UNAVAILABILITY) {
                throw new InternetException($this->errorString, $this->errorNumber);
            }

            return false;
        } catch (InternetException $e) {
            throw $e;
        } catch (Exception $e) {
            if ($this->logger !== null) {
                $this->logger->debug('Something went wrong while trying to check internet availability: ' . $e->getMessage());
            }

            if ($this->options & self::PROBLEM_AS_EXCEPTION) {
                throw new InternetProblemException($this->errorString, $this->errorNumber);
            }

            return (bool) ($this->options & self::PROBLEM_AS_TRUE);
        }
    }

    /**
     * Convenience method with all default settings.
     *
     * @param bool $problemAsTrue
     * @return bool
     */
    public static function available(bool $problemAsTrue = true): bool
    {
        try {
            return (new self())->check();
        } catch (Exception $e) {
            // Assume something else went wrong.
            return $problemAsTrue;
        }
    }
}
